activerecord/lib/active_record/associations/has_one_association.rb

```diff
@@ -33,8 +33,8 @@ def delete(method = options[:dependent])
             target.destroy
             throw(:abort) unless target.destroyed?
           when :destroy_async
-            primary_key_column = target.class.primary_key.to_sym
-            id = target.public_send(primary_key_column)
+            primary_key_column = (target.class.query_constraints_list || Array(target.class.primary_key)).map(&:to_sym)
+            id = primary_key_column.map { |col| target.public_send(primary_key_column) }
```

Comment:
>just based on the presence of query_constraints_list being defined?

Basically this, at this moment we are using the presence of the `query_constraints_list` or `primary_key` being an array almost like a configuration flag for us to know that the application is opted in into the composite-pk behaviors

>if an app opts into using query_constraints_list they're aware that the signature of that job will change?

That's the plan 👍 

Eventually I can imagine Rails going through a deprecation cycle and making the signature always compsoite-pk-ready but since there was no deprecation cycle for that I don't think we can immediately change the signature.

But even if we can change the signature, unless absolutely necessary, it feels like it would be more appropriate to cause less problems for an application being upgraded at an expense of an internal if-condition